### PR TITLE
Autocomplete: implement stage counter logger

### PR DESCRIPTION
**Pull request by valerybugakov**
- Implements a new autocomplete stage logger, which counts the completion requests that reached different stages of the autocomplete generation pipeline.
- It is very similar to https://github.com/sourcegraph/cody/pull/3070, so I assume that "this event is probably a good candidate to exclude from the Amplitude pipeline" based on Kelsey's comment [here](https://github.com/sourcegraph/cody/pull/3070#issuecomment-2003948272). cc @akalia25 
- Closes https://linear.app/sourcegraph/issue/CODY-2413/%5Bautocomplete-latency%5D-add-feedback-rate-logger
- [Slack thread](https://sourcegraph.slack.com/archives/C0729T2PBV2/p1718627075338329?thread_ts=1718357408.712179&cid=C0729T2PBV2)

## Test plan

- Added unit tests
- Ensured that dispose is called and we flush when VS Code is closed

Screenshot 2024-06-18 at 13 48 13

### CODE REVIEW

File: vscode/src/completions/inline-completion-item-provider.ts [hunk 1]
```diff
@@ -385,6 +386,12 @@ export class InlineCompletionItemProvider
                     isDotComUser: this.config.isDotComUser,
                 })
 
+                // Do not increment the `preFinalCancellationCheck` counter if the result is empty.
+                // We don't have an opportunity to show a completion if it's null.
+                if (result) {
+                    autocompleteFeedbackRateLogger.record('preFinalCancellationCheck')
+                }
```
Comment (valerybugakov) [inline vscode/src/completions/inline-completion-item-provider.ts:393]: @hitesh-1997, let me know if you think we don't need a condition here

File: vscode/src/completions/inline-completion-item-provider.ts [hunk 2]
```diff
@@ -385,6 +386,12 @@ export class InlineCompletionItemProvider
                     isDotComUser: this.config.isDotComUser,
                 })
 
+                // Do not increment the `preFinalCancellationCheck` counter if the result is empty.
+                // We don't have an opportunity to show a completion if it's null.
```
Comment (dominiccooney) [inline vscode/src/completions/inline-completion-item-provider.ts:390]: You use two different phrases, `empty` and `null`, to refer to the same state. It appears that you're making a distinction, which is confusing. Don't use different words unless you want to make a distinction.
```suggestion
                // Do not increment the `preFinalCancellationCheck` counter if the result is empty.
                // We don't have an opportunity to show a completion if it's empty.
```

File: vscode/src/services/autocomplete-feedback-rate-logger.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
+import type * as vscode from 'vscode'
+
+import { telemetryRecorder } from '@sourcegraph/cody-shared'
+
+export const LOG_INTERVAL = 30 * 60 * 1000 // 30 minutes
+
+/**
+ * Keys represent autocomplete pipeline stages in chronological order.
+ */
+const INITIAL_STATE = {
```
Comment (dominiccooney) [inline vscode/src/services/autocomplete-feedback-rate-logger.ts:10]: Very succinct typing here with this flowing into `currentState` and then using `keyof typeof this.currentState`. Neat.

File: vscode/src/services/autocomplete-feedback-rate-logger.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    preLastCandidate: 0,
+    preCache: 0,
+    preDebounce: 0,
+    preContextRetrieval: 0,
+    preNetworkRequest: 0,
+    preFinalCancellationCheck: 0,
+    preVisibilityCheck: 0,
+}
+
+export class AutocompleteFeedbackRateLogger implements vscode.Disposable {
```
Comment (dominiccooney) [inline vscode/src/services/autocomplete-feedback-rate-logger.ts:20]: This isn't really a *rate* but counts... `AutocompleteStepCounter`? 🚶 ⌚

**Review by dominiccooney — APPROVED** [review]
Beautiful.

File: vscode/src/services/autocomplete-feedback-rate-logger.ts (hunk 4 above)
Comment (valerybugakov) [reply vscode/src/services/autocomplete-feedback-rate-logger.ts:20]: Agreed, we do not calculate the rate in this logger, but we plan to use these counts to calculate [the feedback rate](https://codeium.com/blog/golden-metrics-characters-per-opportunity-percentage-code-written#:~:text=not%20to%20try.-,Feedback%20Rate,-%3A%20There%20is%20obvious) later. Will update 👍

**Comment by akalia25** [comment]
Thanks for the heads-up @valerybugakov, updated our Amplitude pipeline to not ingest this event. 

CC: @kelsey-brown

**Review by hitesh-1997 — APPROVED** [review]
Thanks for adding this change Valery !!

File: vscode/src/completions/inline-completion-item-provider.ts (hunk 1 above)
Comment (hitesh-1997) [reply vscode/src/completions/inline-completion-item-provider.ts]: should be okay to have i think

**Pull request merged**